File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.java
class KthLargest {
    
    PriorityQueue<Integer> upper = new PriorityQueue<Integer>();
    PriorityQueue<Integer> lower = new PriorityQueue<Integer>();
    int kGlobal = 0;

    public KthLargest(int k, int[] nums) {
        for(int next : nums) {
            upper.add(next);
        }
        
        // leave k largest elements in offer
        kGlobal = k;
        while(upper.size() > k) {
            lower.add(upper.poll());
        }
    }
    
    public int add(int val) {
        if (upper.size() < kGlobal) {
            upper.add(val);
            return upper.peek();
        }
        if (val <= upper.peek()) {
            lower.add(val);
            return upper.peek();
        }
        else {
            lower.add(upper.poll());
            upper.add(val);
            return upper.peek();
        }
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.add(val);
 */

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.kt
class KthLargest(k: Int, nums: IntArray) {
    
    val lower = PriorityQueue<Int>()
    val upper = PriorityQueue<Int>()
    val kSize = k
    
    init {
        for (i in 0 until nums.size) {
            upper.add(nums[i])
        }
        
        while(upper.size > k) {
            lower.add(upper.poll())
        }
    }


    fun add(value: Int): Int {
        if (upper.size < kSize) {
            upper.add(value)
            return upper.peek()
        }
        if (value <= upper.peek()) {
            lower.add(value)
            return upper.peek()
        }
        lower.add(upper.poll())
        upper.add(value)
        return upper.peek()
    }

}

/**
 * Your KthLargest object will be instantiated and called as such:
 * var obj = KthLargest(k, nums)
 * var param_1 = obj.add(`val`)
 */